### PR TITLE
feat: make LLVM eDSL work without reduce - two/three instruction case

### PR DESCRIPTION
**Pull request by tobiasgrosser**
The use of reduce in the meta programming framework has made the frontend pretty fragile. We build now step-by-step the simp lemmas needed to simplify the eDSL away such that reduce can eventually be dropped.

This is likely enough to generalize.

This change is needed to enable the lean 4.9.0 update.

Co-authored-by: Alex Keizer <alex@keizer.dev>
Co-authored-by: Siddharth Bhat <siddu.druid@gmail.com>

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
This is now ready to review.

File: SSA/Core/ErasedContext.lean [hunk 1]
```diff
@@ -143,6 +143,10 @@ def castCtxt {Γ : Ctxt Op} (h_eq : Γ = Δ) : Γ.Var ty → Δ.Var ty
 @[simp]
 theorem toMap_last {Γ : Ctxt Ty} {t : Ty} : (Ctxt.Var.last Γ t).toMap = Ctxt.Var.last (Γ.map f) (f t) := rfl
 
+@[simp]
+theorem toSnoc_toMap {Γ : Ctxt Ty} {t : Ty } {var : Ctxt.Var Γ t} {f : Ty → Ty2} :
+    @Ctxt.Var.toMap _ (Γ.snoc t) _ _ f var.toSnoc = var.toMap.toSnoc := rfl
+
```
Comment (bollu) [inline SSA/Core/ErasedContext.lean:149]:
```suggestion
@[simp]
theorem toSnoc_toMap {Γ : Ctxt Ty} {t : Ty} {var : Ctxt.Var Γ t} {f : Ty → Ty₂} :
    @Ctxt.Var.toMap _ (Γ.snoc t) _ _ f var.toSnoc = var.toMap.toSnoc := rfl

```

File: SSA/Core/Framework.lean [hunk 2]
```diff
@@ -1368,6 +1368,9 @@ section Lemmas
     (Com.lete e body).changeDialect f = Com.lete (e.changeDialect f) (body.changeDialect f) := by
   simp only [List.map_eq_map, changeDialect]
 
+@[simp] lemma HVector.changeDialect_nil (f : DialectMorphism d d') :
+    HVector.changeDialect (eff := eff) f nil = nil := rfl
+
```
Comment (bollu) [inline SSA/Core/Framework.lean:1373]:
```suggestion
@[simp] lemma HVector.changeDialect_nil {eff : EffectKind} (f : DialectMorphism d d') :
    HVector.changeDialect (eff := eff) f nil = nil := rfl

```

File: SSA/Core/ErasedContext.lean (hunk 1 above)
Comment (bollu) [reply SSA/Core/ErasedContext.lean:149]: I would like to see if we can get rid of the `@` based application --- do we really need it?

**Review by bollu — COMMENTED** [review]
LGTM, I'm not a huge fan of `@` ing and then leaving many arguments blank. Can we rewrite it in member notation?

File: SSA/Core/ErasedContext.lean (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Core/ErasedContext.lean]: I shortened it to

```lean
theorem toSnoc_toMap {Γ : Ctxt Ty} {t : Ty} {var : Ctxt.Var Γ t} {f : Ty → Ty₂} :
    var.toSnoc.toMap (Γ := Γ.snoc t) (f := f) = var.toMap.toSnoc := rfl
```

**Pull request merged**